### PR TITLE
Polish the Background Image settings

### LINKED ISSUE #8763: Polish Settings UI: Profile > Background Image
- ~What does "uniform to fill" mean?~ moved to #8769 (individual radio button tooltips)
- Feels cluttered when there is no background image. We should hide/disable the other background image settings until you set a path to the image.

<hr>

MG edit: The following from #8761 feels related
> - "backgroundImage" "desktopWallpaper" special value:
>   - This could probably be better represented with a checkbox labelled "use desktop wallpaper"
>   - When checked, the text box and browse button are disabled.

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

Two parts:
* Hide the BG image settings when no image is specified
* Add a checkbox for "Use desktop wallpaper". When that's checked, the BG image path input is hidden. Unchecking that box restores the path to what it was before.

## PR Checklist
* [x] Closes #8763
* [x] I work here

## Validation Steps Performed
Tested manually

### CODE REVIEW

File: src/cascadia/inc/cppwinrt_utils.h [hunk 1]
```diff
@@ -185,3 +185,27 @@ std::vector<wil::com_ptr<T>> SafeArrayToOwningVector(SAFEARRAY* safeArray)
 
     return result;
 }
+
+#define DECLARE_CONVERTER(nameSpace, className)                                                                   \
```
Comment (carlos-zamora) [inline src/cascadia/inc/cppwinrt_utils.h:189]: Might as well use `DECLARE_CONVERTER` on all the other converter header files. For consistency.

File: src/cascadia/TerminalSettingsEditor/Profiles.cpp [hunk 2]
```diff
@@ -26,6 +26,33 @@ static const std::array<winrt::guid, 2> InBoxProfileGuids{
...
+        _profile{ profile }
+    {
+        // Add a property changed handler to our own property changed event.
+        // When the BackgroundImagePath changes, we _also_ need to change the
+        // value of UseDesktopBGImage.
+        //
+        // We need to do this so if someone manually types "desktopWallpaper"
+        // into the path TextBox, we properly update the checkbox and stored
+        // _lastBgImagePath. Without this, then we'll permanently hide the text
+        // box, prevent it from ever being changed again.
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsEditor/Profiles.cpp:39]: Good catch!

File: src/cascadia/TerminalSettingsEditor/Profiles.cpp [hunk 3]
```diff
@@ -26,6 +26,33 @@ static const std::array<winrt::guid, 2> InBoxProfileGuids{
...
+            if (args.PropertyName() == L"BackgroundImagePath")
+            {
+                _NotifyChanges(L"UseDesktopBGImage");
+            }
+        });
+
+        // Cache the original BG image path. If the user clicks "Use desktop
+        // wallpaper", then un-checks it, this is the string we'll restore to
+        // them.
+        if (BackgroundImagePath() != L"desktopWallpaper")
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsEditor/Profiles.cpp:50]: nit: should `L"desktopWallpaper"` be a `static constexpr std::wstring_view`?

Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsEditor/Profiles.cpp:50]: That way you just reference it below in `UseDesktopBGImage()` too

File: src/cascadia/TerminalSettingsEditor/Profiles.idl [hunk 4]
```diff
@@ -12,6 +12,7 @@ namespace Microsoft.Terminal.Settings.Editor
     {
         Boolean CanDeleteProfile { get; };
         Boolean IsBaseLayer;
+        Boolean UseDesktopBGImage { get; set; };
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsEditor/Profiles.idl:15]:
```suggestion
        Boolean UseDesktopBGImage;
```
Absence of `{ get; set; }` assumes it's there, right?

**Review by carlos-zamora — APPROVED** [review]
Mind adding some pics or a gif to see what it looks like? I think I can visualize it but it'd be nice to have in the PR body.

Love the changes to converters!
Mostly nits here.

**Comment by zadjii-msft** [comment]
sui-desktopWallpaper

**Review by carlos-zamora — CHANGES_REQUESTED** [review]
Realized something. Make sure to disable the "hide background image settings" behavior in the "Base Layer" page. There's a good use case for customizing those settings in "Base Layer" even though you don't have a background image set.

Shouldn't be too difficult. There's an `IsBaseLayer` bool exposed off of the `ProfileViewModel` that you can use to check this.

**Comment by PankajBhojwani** [comment]
I think I would prefer if the checkbox was immediately below the textbox for the path, and that the textbox and button are _disabled_ when the checkbox is set (rather than hidden), this gets us

- consistency with #8786 
- no need to have two 'background image' headers like there currently is, we could just have one
- imo its more intuitive - currently, if the user has 'desktop wallpaper' set and wants to change it, its not immediately obvious that they need to uncheck desktop wallpaper before the textbox will show. But if the textbox is greyed out when the checkbox right underneath it is checked, it becomes more obvious

Thoughts?

**Comment by zadjii-msft** [comment]
image

**Comment by zadjii-msft** [comment]
> 
> 
> Realized something. Make sure to disable the "hide background image settings" behavior in the "Base Layer" page. There's a good use case for customizing those settings in "Base Layer" even though you don't have a background image set.
> 
> Shouldn't be too difficult. There's an `IsBaseLayer` bool exposed off of the `ProfileViewModel` that you can use to check this.

image

**Review by carlos-zamora — APPROVED** [review]
Thanks!

**Review by DHowett — APPROVED** [review]

**Pull request merged**